litejpeg/core/rle/entropycoder.py

```diff
@@ -0,0 +1,223 @@
+"""
+EntropyCoder Module:
+----------------
+This module is been made in order to calculate the number of bits to
+store the amplitude used in the huffman encoding.
+"""
+
+# Importing libraries.
+from litex.gen import *
+from litex.soc.interconnect.stream import *
+
+from litejpeg.core.common import *
+
+
+# To provide delay so in order to sink the data coming from the main
+# module to that of the Datapath module.
+datapath_latency = 3
+
+
+@CEInserter()
+class EntropyDatapath(Module):
+    """
+    EntropyDatapath :
+    ------------------
+    It contains the steps for the EntropyCoder to calculate the bits for the
+    Amplitude to store.
+
+    Attributes:
+    -----------
+    sink : Get the input from the Entrophycoder.
+    source : Give the output to the Entrophycoder.
+    input_data : for temporary storing the value of input, used
+                 for calculating size.
+    get_data : Contains the temporary data for calculating the size.
+    size : The number of bits required for storing th amplitude.
+
+    """
+    def __init__(self):
+
+        # Record the input and output of the Datapath.
+        # sink = input
+        # source = output
+        self.sink = sink = Record(block_layout(12))
+        self.source = source = Record(block_layout(4))
+
+        # Getting the input.
+        input_data = Signal(12)
+
+        # The values for storing the temporary values for
+        # calculating the size.
+        get_data = Array(Signal(12) for i in range(12))
+
+        # Store the output.
+        size = Signal(4)
+
+        # For calculating the size.
+
+        # Storing the size in the input_data.
+        self.sync += input_data.eq(sink.data)
+        # Keep on interating over the input_data and see after shifting
+        # how many bits the input_data value becomes zeros and that is the
+        # number of bits required to store the input_data.
+        for i in range(12):
+            self.sync += get_data[i].eq(input_data >> i)
+        for i in range(12):
```

Comment:
```python
for i in range(11, -1, -1):
```

Reply:
I don't not able to understand this one.

Reply:
@mithro

Reply:
http://pythoncentral.io/pythons-range-function-explained/

Reply:
DONE